clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -151,79 +156,97 @@ public HubConnectionState getConnectionState() {
-                    if (url.contains("?")) {
-                        url = url + "&id=" + negotiateResponse.getConnectionId();
-                    } else {
-                        url = url + "?id=" + negotiateResponse.getConnectionId();
-                    }
-                }
 
-                if (negotiateResponse.getAccessToken() != null) {
-                    this.headers.put("Authorization", "Bearer " + negotiateResponse.getAccessToken());
+    public CompletableFuture start() throws Exception {
+        hubConnectionStateLock.lock();
+        try {
+            if (connectionState != HubConnectionState.DISCONNECTED) {
+                return CompletableFuture.completedFuture(null);
+            }
+            if (!skipNegotiate) {
+                int negotiateAttempts = 0;
+                do {
+                    accessToken = (negotiateResponse == null) ? null : negotiateResponse.getAccessToken();
+                    negotiateResponse = handleNegotiate();
+                    negotiateAttempts++;
+                } while (negotiateResponse.getRedirectUrl() != null && negotiateAttempts < MAX_NEGOTIATE_ATTEMPTS);
+                if (!negotiateResponse.getAvailableTransports().contains("WebSockets")) {
+                    throw new HubException("There were no compatible transports on the server.");
                 }
+            }
 
-                if (negotiateResponse.getRedirectUrl() != null) {
-                    url = this.negotiateResponse.getRedirectUrl();
-                }
+            logger.log(LogLevel.Debug, "Starting HubConnection");
+            if (transport == null) {
+                transport = new WebSocketTransport(url, logger, headers);
+            }
+
+            transport.setOnReceive(this.callback);
+
+            return transport.start().thenCompose((future) -> {
+                String handshake = HandshakeProtocol.createHandshakeRequestMessage(new HandshakeRequestMessage(protocol.getName(), protocol.getVersion()));
+                CompletableFuture handshakeFuture= transport.send(handshake);
+                return handshakeFuture.thenRun(() -> {
+                    connectionState = HubConnectionState.CONNECTED;
+                    logger.log(LogLevel.Information, "HubConnected started.");
+                });
+            });
+        } finally {
+            hubConnectionStateLock.unlock();
+        }
+    }
 
-                negotiateAttempts++;
-            } while (negotiateResponse.getRedirectUrl() != null && negotiateAttempts < MAX_NEGOTIATE_ATTEMPTS);
-            if (!negotiateResponse.getAvailableTransports().contains("WebSockets")) {
-                throw new HubException("There were no compatible transports on the server.");
+    private NegotiateResponse handleNegotiate() throws IOException {
+        accessToken = (negotiateResponse == null) ? null : negotiateResponse.getAccessToken();
+        negotiateResponse = Negotiate.processNegotiate(url, accessToken);
```

Comment:
Shouldn't we make negotiate async in this PR too?

Reply:
I don't want to block invoke any longer than it needs to be 😄